projects/user_publication/gscholar.py

```diff
@@ -0,0 +1,207 @@
+import logging
+import re
+from datetime import datetime
+from pathlib import Path
+
+from scholarly import ProxyGenerator, scholarly
+from scholarly._proxy_generator import MaxTriesExceededException
+
+from projects.models import ChameleonPublication, Publication
+from projects.user_publication import utils
+
+logger = logging.getLogger("projects")
+
+MAX_RETRIES = 10
+
+
+def handle_scrapper():
+    scraper_config = Path("scraper.config")
+    scraper_api = ''
+    if scraper_config.exists():
+        with scraper_config.open('r') as s_config:
+            logger.info("Using scraper.config file for API key")
+            scraper_api = s_config.read()
+    if not scraper_api:
+        scraper_api = input("Enter the scraper API key, if you have one, else press enter: ")
+    if scraper_api:
+        with scraper_config.open('w') as s_config:
+            s_config.write(scraper_api)
+    return scraper_api
+
+
+class GoogleScholarHandler(object):
+    def __init__(self):
+        """To make google scholar calls using proxy"""
+        self.pg = ProxyGenerator()
+        self.scholarly = scholarly
+        scraper_api = handle_scrapper()
+        if scraper_api:
+            self.pg.ScraperAPI(scraper_api)
+            logger.info("Using scrapper proxy")
+        else:
+            self.pg.FreeProxies()
+            logger.info("Using Free proxy")
+        self.scholarly.use_proxy(self.pg)
+        self.retries = 0
+        self.citations = []
+
+    def _handle_proxy_reload(func):
+        def inner_f(self, *args, **kwargs):
+            while self.retries < MAX_RETRIES:

```

Comment:
If it exceeds maximum retries shouldn't it be exiting raising an error?

Reply:
Nothing seems to guarantee that the exception is thrown. This while loop seems to count to 10 and then always return a symbol called `resp`. What happens if it counts all the way to 10? If resp doesn't exist, it will throw a `NameError` which is not a clear explanation of what went wrong. I believe this code is also meant to be run manually, so even if it throws, I think the object owning this method will still exist, no? If I try to run the same code again, it will never work because retries has not been reset to 0.